@dvsav Could you let me know which transformers version did you use?

I've got error with multiple transformers versions.

### 4.57.0

```bash
python tico/quantization/wrapq/examples/qwen/quantize_model.py
┌───────────── Quantization Error Summary ─────────────
│ Mean |diff|: 0.058264
│ PEIR       : 5.270459 %
└──────────────────────────────────────────────────────
    ┌────────────────────────────────────────────┐
 3.8┤                                            │
    │                                         •  │
    │                                      •     │
 2.5┤                                   ••••     │
    │                                 ••••       │
    │                               ••••         │
    │                             ••••           │
 1.3┤                           ••••             │
    │                        •••••               │
    │                      •••••                 │
-0.0┤                    ••••••                  │
    │                   •••••                    │
    │                •••••                       │
    │               •••••                        │
-1.3┤             ••••                           │
    │          •••••                             │
    │         ••••                               │
-2.6┤       ••••                                 │
    │     •••                                    │
    │   •••                                      │
    │   •                                        │
-3.9┤                                            │
    └┬──────────┬──────────┬─────────┬──────────┬┘
   -3.9       -2.0       -0.0       1.9       3.8 

Traceback (most recent call last):
  File "/home/seongwoo/TICO/tico/quantization/wrapq/examples/qwen/quantize_model.py", line 363, in <module>
    main()
  File "/home/seongwoo/TICO/tico/quantization/wrapq/examples/qwen/quantize_model.py", line 354, in main
    circle_model = tico.convert(quantized_model.eval(), example_input)
  File "/home/seongwoo/TICO/tico/utils/convert.py", line 346, in convert
    exported_program = export(
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/torch/export/__init__.py", line 360, in export
    return _export(
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/torch/export/_trace.py", line 1100, in wrapper
    raise e
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/torch/export/_trace.py", line 1066, in wrapper
    ep = fn(*args, **kwargs)
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/torch/export/exported_program.py", line 121, in wrapper
    return fn(*args, **kwargs)
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/torch/export/_trace.py", line 2120, in _export
    ep = _export_for_training(
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/torch/export/_trace.py", line 1100, in wrapper
    raise e
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/torch/export/_trace.py", line 1066, in wrapper
    ep = fn(*args, **kwargs)
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/torch/export/exported_program.py", line 121, in wrapper
    return fn(*args, **kwargs)
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/torch/export/_trace.py", line 1983, in _export_for_training
    export_artifact = export_func(  # type: ignore[operator]
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/torch/export/_trace.py", line 1352, in _strict_export_lower_to_aten_ir
    gm_torch_level = _export_to_torch_ir(
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/torch/export/_trace.py", line 740, in _export_to_torch_ir
    gm_torch_level, _ = torch._dynamo.export(
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 1821, in inner
    graph = rewrite_signature(
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 1355, in rewrite_signature
    matched_output_elements_positions = produce_matching(
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 1341, in produce_matching
    raise AssertionError(
AssertionError: Unexpectedly found a <class 'torch.Tensor'> in the outputs.
Please file an issue along with a paste of the logs from TORCH_LOGS="+export"
```

### 5.5.4

```bash
python tico/quantization/wrapq/examples/qwen/quantize_model.py
Traceback (most recent call last):
  File "/home/seongwoo/TICO/tico/quantization/wrapq/examples/qwen/quantize_model.py", line 363, in <module>
    main()
  File "/home/seongwoo/TICO/tico/quantization/wrapq/examples/qwen/quantize_model.py", line 343, in main
    fp_out = orig_model(**test_input).last_hidden_state
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1751, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1762, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/transformers/utils/generic.py", line 876, in wrapper
    output = func(self, *args, **kwargs)
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/transformers/models/qwen3_vl/modeling_qwen3_vl.py", line 1336, in forward
    position_ids = self.compute_3d_position_ids(
  File "/home/seongwoo/TICO/.venv/lib/python3.10/site-packages/transformers/models/qwen3_vl/modeling_qwen3_vl.py", line 1220, in compute_3d_position_ids
    raise ValueError(
ValueError: Multimodal data was passed (via `image_grid_thw` or `video_grid_thw`) but `mm_token_type_ids` is missing. Please pass `mm_token_type_ids` to the model so that multimodal RoPE (M-RoPE) can be computed correctly. `mm_token_type_ids` is returned by the processor alongside `input_ids`.
```

--- a/test/quantization/wrapq/wrappers/qwen_vl/test_quant_text_model.py
+++ b/test/quantization/wrapq/wrappers/qwen_vl/test_quant_text_model.py
@@ -67,6 +67,41 @@ class TestQuantQwen3VLTextModel(unittest.TestCase):
         cls.hidden_size = cfg.hidden_size
         cls.num_hidden_layers = cfg.num_hidden_layers
 
+    @staticmethod
+    def _create_visual_inputs(
+        vocab_size: int,
+        visual_start_idx: int,
+        grid_thw: tuple[int, int, int],
+        image_token_id: int,
+        vision_start_token_id: int,
+        batch_size: int = 1,
+        patch_size: int = 16,
+        image_width: int = 128,
+        image_height: int = 96,
+    ):
+        num_visual_tokens = image_width * image_height // (patch_size**2)
+        num_tokens = visual_start_idx + num_visual_tokens + 10
+
+        input_ids = torch.randint(
+            low=0, high=vocab_size - 3, size=(batch_size, num_tokens)
+        )
+        input_ids[:, visual_start_idx - 1] = vision_start_token_id
+        input_ids[
+            :, visual_start_idx : visual_start_idx + num_visual_tokens
+        ] = image_token_id
+        attention_mask = torch.ones(batch_size, num_tokens, dtype=torch.int64)
+        pixel_values = torch.full(
+            size=(batch_size, 3, image_height, image_width), fill_value=-1.0
+        )
+        image_grid_thw = torch.tensor(grid_thw)
+
+        return {
+            "input_ids": input_ids,
+            "attention_mask": attention_mask,
+            "pixel_values": pixel_values,
+            "image_grid_thw": image_grid_thw,
+        }
+
     def _create_test_inputs(self, batch_size=2, seq_len=16):
         """Helper to create test inputs for TextModel."""
         input_ids = torch.randint(0, self.vocab_size, (batch_size, seq_len))
@@ -826,3 +861,51 @@ class TestQuantQwen3VLTextModel(unittest.TestCase):
         self.assertIsInstance(q_out, tuple)
         self.assertEqual(len(q_out), 1)
         self.assertEqual(q_out[0].shape, (1, 8, self.hidden_size))
+
+    def test_forward_diff_export_time(self):
+        """
+        Test that quantized output is acceptably close to FP32 reference at export time.
+        """
+        torch.manual_seed(42)
+        grid_thw = (1, 8, 8)
+        visual_start_idx = 4
+        vocab_size = self.fp_model.config.vocab_size
+
+        ptq_config = PTQConfig(
+            model_args={
+                "vision": {
+                    "grid_thw": grid_thw,
+                    "visual_start_idx": visual_start_idx,
+                }
+            }
+        )
+        q_model = QuantQwen3VLTextModel(self.fp_model, qcfg=ptq_config)
+        q_model.enable_calibration()
+
+        model_inputs: dict = self._create_visual_inputs(
+            vocab_size=vocab_size,
+            visual_start_idx=visual_start_idx,
+            image_token_id=vocab_size - 2,
+            vision_start_token_id=vocab_size - 3,
+            grid_thw=grid_thw,
+            batch_size=1,
+        )
+
+        # Calibrate the model
+        _ = q_model(**model_inputs)
+
+        q_model.freeze_qparams()
+        q_model.force_export = True
+
+        with torch.no_grad():
+            q_out = q_model(**model_inputs)
+            fp_out = self.fp_model(**model_inputs)
+
+        # Extract last_hidden_state
+        q_hidden = q_out.last_hidden_state
+        fp_hidden = fp_out.last_hidden_state
+
+        self.assertEqual(fp_hidden.shape, q_hidden.shape)
+        diff = (fp_hidden - q_hidden).abs().mean().item()
+        self.assertGreater(diff, 0.0)  # not identical
+        self.assertLess(diff, 0.7)  # acceptably close

--- a/tico/quantization/wrapq/examples/qwen/quantize_for_conditional_generation.py
+++ b/tico/quantization/wrapq/examples/qwen/quantize_for_conditional_generation.py
@@ -301,6 +301,7 @@ def main():
         model_args={
             "vision": {
                 "grid_thw": thw,
+                "visual_start_idx": 0,
             }
         }
     )

--- a/tico/quantization/wrapq/examples/qwen/quantize_model.py
+++ b/tico/quantization/wrapq/examples/qwen/quantize_model.py
@@ -82,6 +82,7 @@ def create_visual_input(
     spatial_patch_size: int,
     vocab_size: int,
     image_token_id: int,
+    visual_start_idx: int,
 ):
     """Helper to create input with videos or images."""
     assert (
@@ -109,7 +110,9 @@ def create_visual_input(
     # Replace first tokens with video placeholder tokens
     # This marks where the video features should be inserted
     for i in range(batch_size):
-        input_ids[i, :num_video_tokens] = image_token_id
+        input_ids[
+            i, visual_start_idx : visual_start_idx + num_video_tokens
+        ] = image_token_id
 
     num_temporal_patches, num_spatial_patches_h, num_spatial_patches_w = thw
 
@@ -244,6 +247,7 @@ def generate_calibration_data(
     spatial_patch_size: int,
     vocab_size: int,
     image_token_id: int,
+    visual_start_idx: int,
 ):
     calibration_data = []
     for i in range(batch_size):
@@ -255,6 +259,7 @@ def generate_calibration_data(
             spatial_patch_size,
             vocab_size,
             image_token_id,
+            visual_start_idx=visual_start_idx,
         )
         calibration_data.append(x)
     return calibration_data
@@ -289,12 +294,14 @@ def main():
         video_token_id=999,
     )
     thw = (1, 8, 8)
+    visual_start_idx = 0
 
     # Configure PTQ
     ptq_config = tico.quantization.config.ptq.PTQConfig(
         model_args={
             "vision": {
                 "grid_thw": thw,
+                "visual_start_idx": visual_start_idx,
             }
         }
     )
@@ -319,6 +326,7 @@ def main():
         spatial_patch_size=cfg.vision_config.patch_size,
         vocab_size=cfg.text_config.vocab_size,
         image_token_id=cfg.image_token_id,
+        visual_start_idx=visual_start_idx,
     )
 
     # Calibrate the model (collect statistics)

--- a/tico/quantization/wrapq/examples/qwen/trace_qwen.py
+++ b/tico/quantization/wrapq/examples/qwen/trace_qwen.py
@@ -228,6 +228,7 @@ def build_vlm_inputs(
 
 def prepare_inputs(
     image_token_id: int,
+    vision_start_token_id: int,
     vocab_size: int,
     model_name: ModelNameOrPath,
     cache_dir: DirPath | None = None,
@@ -278,9 +279,11 @@ def prepare_inputs(
     )
 
     # Normalize input_ids to be consistent with our image_token_id
-    image_pad_token_id = 151655
+    old_image_pad_token_id = 151655
+    old_vision_start_token_id = 151652
     input_ids: torch.Tensor = model_inputs["input_ids"]
-    input_ids[input_ids == image_pad_token_id] = image_token_id
+    input_ids[input_ids == old_image_pad_token_id] = image_token_id
+    input_ids[input_ids == old_vision_start_token_id] = vision_start_token_id
 
     # Make sure that our input IDs don't go beyond vocabulary size
     input_ids = input_ids % vocab_size
@@ -318,6 +321,7 @@ def prepare_config() -> Qwen3VLConfig:
     Returns:
         Qwen3VLConfig with reduced sizes suitable for quick testing.
     """
+    vocab_size = 1000
 
     cfg = Qwen3VLConfig(
         vision_config={
@@ -339,15 +343,17 @@ def prepare_config() -> Qwen3VLConfig:
             "attention_bias": False,
             "attention_dropout": 0.0,
             "max_position_embeddings": 1024,
-            "vocab_size": 1000,
+            "vocab_size": vocab_size,
             "use_cache": False,
             "rope_scaling": {"rope_type": "default", "mrope_section": [1, 1, 2]},
         },
-        image_token_id=998,
-        video_token_id=999,
+        image_token_id=vocab_size - 2,
+        video_token_id=vocab_size - 1,
+        vision_start_token_id=vocab_size - 3,
     )
     assert cfg.image_token_id < cfg.text_config.vocab_size
     assert cfg.video_token_id < cfg.text_config.vocab_size
+    assert cfg.vision_start_token_id < cfg.text_config.vocab_size
 
     # Ensure eager attention implementation so outputs are deterministic
     # and do not require GPU flash attention kernels.
@@ -385,7 +391,7 @@ def prepare_quantized_model(
         model_args={
             "vision": {
                 "grid_thw": thw,
-                "visual_start_idx": 0,
+                "visual_start_idx": 4,
             }
         },
     )
@@ -1092,6 +1098,7 @@ def main():
         model_name=args.model,
         cache_dir=args.cache_dir,
         image_token_id=cfg.image_token_id,
+        vision_start_token_id=cfg.vision_start_token_id,
         vocab_size=cfg.text_config.vocab_size,
         image_width=128,
         image_height=96,

--- a/tico/quantization/wrapq/wrappers/qwen_vl/quant_text_model.py
+++ b/tico/quantization/wrapq/wrappers/qwen_vl/quant_text_model.py
@@ -25,98 +25,6 @@ from tico.quantization.wrapq.wrappers.registry import try_register
 from transformers.cache_utils import Cache
 
 
-def apply_interleaved_mrope(self, freqs, mrope_section):
-    """
-    Apply interleaved MRoPE to 3D rotary embeddings.
-    Reorganizes frequency layout from chunked [TTT...HHH...WWW] to
-    interleaved [THWTHWTHW...TT], preserving frequency continuity.
-
-    Args:
-        freqs: (3, bs, seq_len, head_dim // 2)
-        mrope_section: (3,)
-
-    Returns:
-        freqs_t: (bs, seq_len, head_dim // 2)
-
-    Design Note:
-        This implementation is using slice_copy, index_select, and cat
-        to avoid yet unsupported slice_scatter with step=3 operation and
-        to avoid unsupported in-place operator index_put.default.
-    """
-    # Start with T dimension (will keep some, replace some)
-    freqs_t_base = freqs[0]
-
-    # For each dimension (H, W), extract frequency bands to be interleaved
-    h_w_bands = []
-
-    for dim, offset in enumerate((1, 2), start=1):  # H, W dimensions
-        length = mrope_section[dim] * 3
-        indices = torch.arange(offset, length, 3, device=freqs.device)
-
-        # Select frequency bands from H/W dimensions
-        # freqs[dim] has shape (bs, seq_len, head_dim//2)
-        # index_select on last dim: (bs, seq_len, num_selected)
-        freqs_bands = freqs[dim].index_select(dim=-1, index=indices)
-        h_w_bands.append(freqs_bands)
-
-    # Now we need to build the interleaved output
-    # Original T dimension indices range from 0 to (head_dim // 2 - 1)
-    # We want to replace specific indices with H/W bands
-
-    # The interleaving pattern: T0, H1, W2, T3, H4, W5, T6, H7, W8, ...
-    # - Positions where i % 3 == 0: T dimension (unchanged from freqs[0])
-    # - Positions where i % 3 == 1: H dimension (overwritten from freqs[1])
-    # - Positions where i % 3 == 2: W dimension (overwritten from freqs[2])
-    # After mrope_section[dim] * 3 positions, H/W bands are exhausted and
-    # fallback to T values for all remaining mod 1 and mod 2 positions.
-
-    # Build the output by slicing and concatenating
-    # Strategy: Slice T dimension into chunks, insert H/W bands, concatenate
-
-    chunks = []
-    pos = 0
-
-    # Total length in the last dimension
-    total_len = freqs_t_base.shape[-1]
-
-    for i in range(total_len):
-        # Determine which dimension this position belongs to
-        # Pattern: T, H, W, T, H, W, T, H, W, ... (repeating cycle)
-        mod = i % 3
-
-        if mod == 0:
-            # T dimension position - take from T
-            # Slice just this index from T
-            chunk = freqs_t_base[..., i : i + 1]
-            chunks.append(chunk)
-        elif mod == 1:
-            # H dimension position - take from H
-            # Calculate which band this is
-            band_idx = (i - 1) // 3
-            if band_idx < h_w_bands[0].shape[-1]:
-                chunk = h_w_bands[0][..., band_idx : band_idx + 1]
-                chunks.append(chunk)
-            else:
-                # Fallback to T if out of bounds
-                chunk = freqs_t_base[..., i : i + 1]
-                chunks.append(chunk)
-        else:  # mod == 2
-            # W dimension position - take from W
-            band_idx = (i - 2) // 3
-            if band_idx < h_w_bands[1].shape[-1]:
-                chunk = h_w_bands[1][..., band_idx : band_idx + 1]
-                chunks.append(chunk)
-            else:
-                # Fallback to T if out of bounds
-                chunk = freqs_t_base[..., i : i + 1]
-                chunks.append(chunk)
-
-    # Concatenate all chunks
-    freqs_t = torch.cat(chunks, dim=-1)
-
-    return freqs_t
-
-
 @try_register(
     "transformers.models.qwen3_vl.modeling_qwen3_vl.Qwen3VLTextModel",
 )
@@ -130,6 +38,10 @@ class QuantQwen3VLTextModel(QuantModuleBase):
     - Final normalization layer (norm)
     - Rotary position embedding (rotary_emb) - NOT wrapped
     """
+
+    # This boolean flag enforces model behavior that is only activated during model graph tracing (torch.export.export).
+    # This flag is used in unit tests only in order to check the behavior without actually exporting the model.
+    force_export: bool = False
 
     def __init__(
         self,
@@ -178,11 +90,8 @@ class QuantQwen3VLTextModel(QuantModuleBase):
 
         # rotary_emb
         self.rotary_emb = fp_model.rotary_emb
-        # The original Qwen3VLTextRotaryEmbedding.apply_interleaved_mrope emits `slice_scatter with step > 1`.
-        # Replace it with pure functional implementation using basic tensor slicing, index_select, and cat.
-        self.rotary_emb.apply_interleaved_mrope = types.MethodType(
-            apply_interleaved_mrope, self.rotary_emb
-        )
+
+        device = next(fp_model.parameters()).device
 
         # ----- static buffers: causal mask template ---------------------------
         assert isinstance(self.config.max_position_embeddings, int)
@@ -190,6 +99,27 @@ class QuantQwen3VLTextModel(QuantModuleBase):
         mask = torch.full((1, 1, max_seq, max_seq), float("-120"))
         mask.triu_(1)
         self.register_buffer("causal_mask_template", mask, persistent=False)
+
+        # ----- static buffers: position_ids -----------------------------------
+        visual_start_idx = self._get_visual_start_idx(qcfg)
+        grid_thw = self._get_vision_grid_thw(qcfg)
+        position_ids = self._compute_3d_position_ids(
+            device=device,
+            visual_start_idx=visual_start_idx,
+            seq_len=max_seq,
+            thw=grid_thw,
+            spatial_merge_size=2,  # WARNING: hard-coded value
+        )
+        self.register_buffer("position_ids_template", position_ids, persistent=False)
+
+        # ----- static buffers: position_embeddings -----------------------------------
+        _ = torch.empty(0, device=device)
+        # tuple of 2 tensors with shape (batch_size, seq_len, head_dim // 2)
+        cos, sin = self.rotary_emb(_, position_ids)
+        assert cos.shape[:-1] == (1, max_seq)
+        assert sin.shape[:-1] == (1, max_seq)
+        self.register_buffer("cos_template", cos, persistent=False)
+        self.register_buffer("sin_template", sin, persistent=False)
 
         # --- Observers for floating-point tensors -----------------------------
         mk = self._make_obs
@@ -388,6 +318,169 @@ class QuantQwen3VLTextModel(QuantModuleBase):
             f"Expected None, 2D, or 4D mask, but got ndim={attention_mask.ndim}."
         )
 
+    @staticmethod
+    def _compute_3d_position_ids(
+        device: torch.device,
+        visual_start_idx: int | None,
+        seq_len: int,
+        thw: tuple[int, int, int] | None,
+        spatial_merge_size: int,
+    ) -> torch.Tensor:
+        """
+        Compute 3D position IDs for multimodal RoPE with simplified constraints.
+
+        This function assumes:
+        - Single sample in batch (batch_size = 1)
+        - Image tokens always begin at a fixed visual_start_idx position
+        - Fixed sequence length
+
+        Parameters:
+        - device: The device to create tensors on
+        - visual_start_idx: The fixed position where image tokens begin
+        - seq_len: The fixed sequence length
+        - thw: Tuple of (temporal, height, width) dimensions for vision tokens
+        - spatial_merge_size: Spatial merge size for vision tokens
+
+        Returns:
+        - position_ids: Tensor of shape (3, 1, seq_len) with 3D position IDs
+        """
+        if thw is None:
+            # Generate simple 1D position IDs for text-only input
+            position_ids = torch.arange(seq_len, device=device).expand(3, 1, seq_len)
+            return position_ids
+
+        assert visual_start_idx is not None
+
+        # Initialize position_ids tensor
+        position_ids = torch.ones(3, 1, seq_len, dtype=torch.long, device=device)
+
+        # List to store position ID segments
+        pos_ids_list: list[torch.Tensor] = []
+
+        # Text position IDs (before visual tokens)
+        if visual_start_idx > 0:
+            text_len = visual_start_idx
+            text_pos_ids = (
+                torch.arange(text_len, device=device).view(1, -1).expand(3, -1)
+            )
+            pos_ids_list.append(text_pos_ids)
+
+        # Vision position IDs (3D)
+        llm_grid_t = thw[0]
+        llm_grid_h = thw[1] // spatial_merge_size or 1
+        llm_grid_w = thw[2] // spatial_merge_size or 1
+
+        # Create 3D position indices
+        t_index = (
+            torch.arange(llm_grid_t, device=device)
+            .view(-1, 1)
+            .expand(-1, llm_grid_h * llm_grid_w)
+            .flatten()
+        )
+        h_index = (
+            torch.arange(llm_grid_h, device=device)
+            .view(1, -1, 1)
+            .expand(llm_grid_t, -1, llm_grid_w)
+            .flatten()
+        )
+        w_index = (
+            torch.arange(llm_grid_w, device=device)
+            .view(1, 1, -1)
+            .expand(llm_grid_t, llm_grid_h, -1)
+            .flatten()
+        )
+
+        # Starting index for vision tokens
+        st_idx = visual_start_idx
+        vision_pos_ids = torch.stack([t_index, h_index, w_index]) + st_idx
+        pos_ids_list.append(vision_pos_ids)
+
+        # Trailing text position IDs (after visual tokens)
+        num_visual_tokens = llm_grid_t * llm_grid_h * llm_grid_w
+        visual_end_idx = visual_start_idx + num_visual_tokens
+
+        if visual_end_idx < seq_len:
+            st_idx = pos_ids_list[-1].max() + 1 if len(pos_ids_list) > 0 else 0
+            trailing_text_len = seq_len - visual_end_idx
+            trailing_pos_ids = (
+                torch.arange(trailing_text_len, device=device).view(1, -1).expand(3, -1)
+                + st_idx
+            )
+            pos_ids_list.append(trailing_pos_ids)
+
+        # Concatenate all position ID segments
+        if pos_ids_list:
+            final_pos_ids = torch.cat(pos_ids_list, dim=1).reshape(3, -1)
+            position_ids[..., 0, :] = final_pos_ids
+
+        return position_ids  # shape = (3, 1, seq_len) where 3=THW, 1=batch_size
+
+    @staticmethod
+    def _get_vision_grid_thw(qcfg: Optional[PTQConfig]) -> torch.Tensor | None:
+        """
+        Extract vision grid shape from PTQConfig.model_args.
+
+        Expected format:
+            PTQConfig(
+                model_args={
+                    "vision": {
+                        "grid_thw": (T, H, W),
+                    }
+                }
+            )
+        """
+        if qcfg is None:
+            return None
+
+        vision_args = qcfg.get_model_arg("vision", {})
+        if not isinstance(vision_args, dict):
+            raise ValueError(
+                "PTQConfig.model_args['vision'] must be a mapping containing "
+                "'grid_thw'."
+            )
+
+        grid_thw = vision_args.get("grid_thw")
+        if grid_thw is not None:
+            assert type(grid_thw) is tuple
+            assert len(grid_thw) == 3
+
+        return grid_thw
+
+    @staticmethod
+    def _get_visual_start_idx(qcfg: Optional[PTQConfig]) -> int | None:
+        """
+        Extract vision tokens starting index in the input prompt
+        from PTQConfig.model_args.
+
+        Expected format:
+            PTQConfig(
+                model_args={
+                    "vision": {
+                        "visual_start_idx": img_start_idx,
+                    }
+                }
+            )
+        """
+        if qcfg is None:
+            return None
+
+        vision_args = qcfg.get_model_arg("vision", {})
+        if not isinstance(vision_args, dict):
+            raise ValueError(
+                "PTQConfig.model_args['vision'] must be a mapping containing "
+                "'visual_start_idx'."
+            )
+
+        visual_start_idx = vision_args.get("visual_start_idx")
+        if visual_start_idx is not None:
+            visual_start_idx = int(visual_start_idx)
+            if visual_start_idx < 0:
+                raise ValueError(
+                    f"vision.visual_start_idx must be greater than zero, but got {visual_start_idx}."
+                )
+
+        return visual_start_idx
+
     def forward(
         self,
         input_ids: torch.LongTensor | None = None,
@@ -440,21 +533,32 @@ class QuantQwen3VLTextModel(QuantModuleBase):
         else:
             inputs_embeds = self._fq(inputs_embeds, self.obs_inputs_embeds)
 
+        batch_size, seq_len, _ = inputs_embeds.shape
+        past_seen_tokens = (
+            past_key_values.get_seq_length() if past_key_values is not None else 0
+        )
+
         # Handle cache_position (integer tensor, not quantized)
         if cache_position is None:
-            past_seen_tokens = self._get_past_seen_tokens(past_key_values)
             cache_position = torch.arange(
                 past_seen_tokens,
-                past_seen_tokens + inputs_embeds.shape[1],
+                past_seen_tokens + seq_len,
                 device=inputs_embeds.device,
             )
 
-        # Handle position_ids (integer tensor, not quantized)
-        # the hard coded `3` is for temporal, height and width.
         if position_ids is None:
-            position_ids = cache_position.view(1, 1, -1).expand(
-                3, inputs_embeds.shape[0], -1
-            )
+            if torch.compiler.is_compiling() or self.force_export:
+                # Use precomputed position IDs at export time only
+                # Get position_ids from precomputed position_ids_template buffer.
+                # Take first seq_len positions. We obtain a tensor of shape (3, 1, seq_len).
+                position_ids = self.position_ids_template[
+                    ..., past_seen_tokens : past_seen_tokens + seq_len
+                ]
+            else:
+                position_ids = cache_position.view(1, 1, -1)
+            # Replicate position_ids across batch dimension
+            position_ids = position_ids.expand(3, batch_size, -1)
+            assert position_ids.shape == (3, batch_size, seq_len)
         elif position_ids.ndim == 2:
             position_ids = position_ids[None, ...].expand(3, position_ids.shape[0], -1)
 
@@ -462,6 +566,7 @@ class QuantQwen3VLTextModel(QuantModuleBase):
             text_position_ids = position_ids[0]
             position_ids = position_ids[1:]
         else:
+            # Take 0-th dimension (T - Temporal) from THW (Temporal, Height, Width)
             text_position_ids = position_ids[0]
 
         attention_mask = self._normalize_attention_mask(
@@ -475,8 +580,21 @@ class QuantQwen3VLTextModel(QuantModuleBase):
 
         # create position embeddings to be shared across the decoder layers
         # rotary_emb returns (cos, sin) which are float tensors and need quantization
-        position_embeddings = self.rotary_emb(hidden_states, position_ids)
-        cos, sin = position_embeddings
+
+        if torch.compiler.is_compiling() or self.force_export:
+            # Use precomputed position embeddings at export time only
+            # Get position embeddings from precomputed position_embeddings_template.
+            # Take seq_len positions starting from past_seen_tokens.
+            # We obtain a tensor of shape (1, seq_len, head_dim // 2)
+            cos = self.cos_template[:, past_seen_tokens : past_seen_tokens + seq_len, :]
+            sin = self.sin_template[:, past_seen_tokens : past_seen_tokens + seq_len, :]
+            # Replicate position_embeddings across batch dimension
+            cos = cos.expand(batch_size, -1, -1)
+            sin = sin.expand(batch_size, -1, -1)
+        else:
+            cos, sin = self.rotary_emb(hidden_states, position_ids)
+        position_embeddings = (cos, sin)
+
         position_embeddings = (
             self._fq(cos, self.obs_cos),
             self._fq(sin, self.obs_sin),

